Dockerfile_incl

```diff
@@ -0,0 +1,62 @@
+# syntax=docker/dockerfile:1
+
+# Dockerfile_incl is built as the `commcarehq_incl` image, and is used
+# for running a simple but complete CommCare HQ environment that
+# includes Pillowtop and Celery containers.
+
+FROM python:3.9
+MAINTAINER Dimagi <devops@dimagi.com>
+
+ENV PYTHONUNBUFFERED=1 \
+    PYTHONUSERBASE=/vendor \
+    PATH=/vendor/bin:$PATH \
+    NODE_VERSION=14.19.1
+
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends \
+        default-jdk \
+        wget \
+        libxml2-dev \
+        libxmlsec1-dev \
+        libxmlsec1-openssl \
+        gettext
+
+# Deletes all package sources, so don't apt-get install anything after this:
+RUN rm -rf /var/lib/apt/lists/* /src/*.deb
+
+# Install Node
+RUN curl -SLO "https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION-linux-x64.tar.gz" && \
+    tar -xzf "node-v$NODE_VERSION-linux-x64.tar.gz" -C /usr/local --strip-components=1 && \
+    rm "node-v$NODE_VERSION-linux-x64.tar.gz"
+
+RUN npm -g install \
+    yarn \
+    bower \
+    grunt-cli \
+    uglify-js
+
+# prefer https for git checkouts made by pip
+RUN git config --global url."https://".insteadOf git://
+
+WORKDIR /vendor
+RUN mkdir /sharedfiles
+RUN groupadd -r cchq && \
+    useradd -r -m -g cchq cchq && \
+    chown cchq:cchq /vendor /sharedfiles
+USER cchq:cchq
+
+#COPY requirements/prod-requirements.txt /vendor/requirements.txt
+# TODO: manage.py appears to need nose. How does prod avoid this?
+COPY requirements/dev-requirements.txt /vendor/requirements.txt
```

Comment:
The `COPY` instructions are performed as `root` inside the container, even these that happen after the `USER` instruction.  This causes a problem on systems where a restrictive umask is enforced (e.g. `umask 0077`) because the `cchq` user will not be able to read any filesystem objects that are created by `COPY` (and not explicitly `chmod`'d and/or `chown`'d).

If the non-privileged user (`cchq` in this case) is allowed to edit these items, then [COPY's chown option](https://docs.docker.com/engine/reference/builder/#copy) can resolve this. Otherwise the items should be adjusted with `RUN chmod ...` instructions to set any required permissions.

Reply:
Thanks!

Reply:
> where a restrictive umask is enforced

:thinking: We don't want the cchq user to be changing the code. In fact, the point of these containers is that they're immutable (if we overlook what we're doing with that `/vendor/logs` directory \</cough>) so they can be swapped out for upgrades.

So wouldn't the solution be to fix the umask, if necessary, rather than change the ownership?

Reply:
> We don't want the cchq user to be changing the code.

This is in line with best practices for running services on *nix systems, so it seems like a good goal.

I'm not sure that you can enforce a umask for the `COPY` instructions without somehow altering the shell where `docker build ...` is executed. Assuming all the _ownerships_ are exactly as you want them right now (i.e. `root` owns most things, `cchq` owns a few things due to explicit `RUN chown ...`), then adding something like this should suffice:

```Dockerfile
RUN chmod -R og+rX <paths>
```

You will need to do this after each `COPY` instruction, or alternatively, move all the `COPY`s _prior_ to any unprivileged `RUN`s and do one big recursive `chmod` on the parent directory[ies].

Reply:
**Update**: Oops, all `chmod`s would have to be done prior to the `USER` instruction.